easybuild/easyblocks/generic/apptainer.py

```diff
@@ -0,0 +1,125 @@
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# http://github.com/hpcugent/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+General EasyBuild support for software, using containerized applications
+
+@author: Maxime Boissonneault (Universitle Laval, Calcul Quebec, Digital Research Alliance of Canada)
+"""
+import os
+import re
+from easybuild.framework.easyblock import EasyBlock
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.modules import get_software_root, get_software_version
+from easybuild.tools.run import run_cmd
+
+
+class Apptainer(EasyBlock):
+    """
+    Support for installing software via a container
+    """
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Extra easyconfig parameters specific to Binary easyblock."""
+        extra_vars = EasyBlock.extra_options(extra_vars)
+        extra_vars.update({
+            'aliases': [[], "Commands to alias in the module.", CUSTOM],
+            'source_param': [None, "Source parameter to pass to the build script.", CUSTOM],
+            'source_type': [None, "Type of source used by the build script", CUSTOM],
+            'apptainer_params': ["", "Default parameters for apptainer", CUSTOM],
+        })
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Initialize custom class variables."""
+        super(Apptainer, self).__init__(*args, **kwargs)
+        self.containerpath = os.path.join("/cvmfs/containers.computecanada.ca/content/containers", "-".join([self.name.lower(), self.version]))
+
+    def configure_step(self):
+        """No configuration, included in the container recipe"""
+        pass
+
+    def build_step(self):
+        """Compilation done in install step"""
+        pass
+
+    def install_step(self):
+        """Build and install the container."""
+        cmd = "/bin/sudo -iu containeruser build_container_image.sh -t sandbox -v %s -n %s -s %s -i %s" % (self.version, self.name.lower(), self.cfg['source_param'], self.cfg['source_type'])
```

Comment:
this could be configurable as well

Reply:
configurable how ? 

short of having a parameter `run_this_command_to_install` with whatever people want to put in there (which I don't think we should do)

Reply:
I think the way to make this configurable is to code the whole building of the container within the EasyBlock and not use an external script, but that is a much larger project that I'm not willing to undertake unless there is some serious interest from upstream.

Reply:
You can make this use the `binary.py` easyblock which has
` cmd = ' '.join([self.cfg['preinstallopts'], install_cmd, self.cfg['installopts']]`
then put the info in the easyconfig like this:
```
install_cmd = "/bin/sudo -iu containeruser build_container_image.sh"
installopts = "-v %(version)s -n %(namelower)s -s ... -i... "
```
Now if you prefer not to copy `install_cmd` in all easyconfigs that use this, I'd just put it as a constant at the top of the easyblock for now.

Reply:
ok, converted to the Binary EasyBlock, but with a default command that is not changeable (at least until we get better).